Clarify risks around ordering terms aggregation (#86528)
elastic/elasticsearch: docs/reference/aggregations/bucket/terms-aggregation.asciidoc

[[search-aggregations-bucket-terms-aggregation-order]]
==== Order

By default, the `terms` aggregation orders terms by descending document
`_count`.  This produces bounded error, which elasticsearch can report (see

> <<terms-agg-doc-count-error>>.

Reviewer: missing right parenthesis.

> `_count`.  This produces bounded error, which elasticsearch can report (see
> <<terms-agg-doc-count-error>>.)

Reviewer 2: ```suggestion
`_count`.  This produces a bounded <<terms-agg-doc-count-error,document count>>
error that {es} can report.
```

You can use the `order` parameter to specify a different sort order.  It is extremely
easy to create a terms ordering that will just return wrong results, and not

> obvious to see when you have done so.  Change this only with caution.

Reviewer 2: Do we want to discourage use of the `order` parameter altogether? Given that it's very easy to create a terms ordering that returns inaccurate results, I wonder if we want to expand on the **WARNING** message on line 219 and just say "don't use this". 

WARNING: Avoid using the `order` parameter to specify a different sort order. Due to the way
the `terms` aggregation
<<search-aggregations-bucket-terms-aggregation-shard-size,gets terms from
shards>>, sorting by ascending doc count often produces inaccurate results. If you need to
find rare terms, use the <<search-aggregations-bucket-rare-terms-aggregation,`rare_terms`>>
aggregation instead.

> obvious to see when you have done so.  Change this only with caution.

Reviewer 2: Do we want to discourage use of the `order` parameter altogether? Given that it's very easy to create a terms ordering that returns inaccurate results, I wonder if we want to expand on the **WARNING** message on line 219 and just say "don't use this". 

> WARNING: Avoid using the `order` parameter to specify a different sort order. Due to the way
> the `terms` aggregation
> <<search-aggregations-bucket-terms-aggregation-shard-size,gets terms from
> shards>>, sorting by ascending doc count often produces inaccurate results. If you need to
> find rare terms, use the <<search-aggregations-bucket-rare-terms-aggregation,`rare_terms`>>
> aggregation instead.

Reviewer 2: Thanks for the context @not-napoleon. I think your instincts are correct here. We might want to add a little more context about why it's so easy to create a terms ordering that won't return correct results. What can we replace as the [reason] in the following sentence? Is this easily explained?

> You can use the `order` parameter to specify a different sort order, but we don't recommend it. Because [reason], it's easy to create a terms ordering that returns incorrect results. Use this parameter with caution.

<<search-aggregations-bucket-terms-aggregation-shard-size,gets terms from
shards>>, sorting by ascending doc count often produces inaccurate results.

===== Ordering by the term value

> In this case, the buckets are ordered by the actual term values, e.g.
> Lexicographically for keywords or numerically for numbers.  This sorting is
> safe in both ascending and descending directions, and will always produce
> correct results.

Reviewer 2: ```suggestion
In this case, the buckets are ordered by the actual term values, such as
lexicographic order for keywords or numerically for numbers. This sorting is
safe in both ascending and descending directions, and produces accurate
results.
```

Reviewer 2: We typically want to avoid phrases that include "always", which might not always be true 💡 . Instead, I think we can say that this sorting "produces accurate results".

Author: Nik and I spent like 20 minutes proving it before I typed "always", but yeah, as Knuth said "be careful, I've only proven it correct, I haven't actually tested it".

}
--------------------------------------------------

===== Ordering by a sub aggregation

WARNING: Sorting by a sub aggregation generally produces incorrect ordering, due to the way the `terms` aggregation
<<search-aggregations-bucket-terms-aggregation-shard-size,gets results from
shards>>.

> There are two cases when sub-aggregation sorting is safe.  Sorting by a maximum
> in descending order, or a minimum in ascending order will be correct.  This
> works only because the sort order aligns with the behavior of the sub
> aggregations, i.e. sorting maxes by largest value first or mins by lowest value
> first.

Reviewer 2: ```suggestion
There are two cases when sub-aggregation sorting is safe and returns correct results:
sorting by a maximum in descending order, or a sorting by a minimum in ascending order. 
Because the sort order aligns with the behavior of the sub aggregations, you can sort
`max` buckets by largest value first or `min` buckets by lowest value first.

```

Reviewer 2: I think that you meant `max` and `min` buckets, but I might have misinterpreted your intention. Let me know if I'm off base here.

Reviewer 2: Tricky, indeed! Here's another rendering that hopefully is more accurate:

> There are two cases when sub-aggregation ordering is safe and returns correct results: sorting by a maximum in descending order, or sorting by a minimum in ascending order. These approaches work because they align with the behavior of sub aggregations. That is, if you're looking for the largest maximum or the smallest minimum, the global answer (from combined shards) must be included in one of the local shard answers. Conversely, the smallest maximum and largest minimum wouldn't be accurately computed.

Author: Thanks, that reads pretty well to me.  I'll make the change.

NOTE: In the event that two buckets share the same values for all order criteria the bucket's term value is used as a
tie-breaker in ascending alphabetical order to prevent non-deterministic ordering of buckets.

===== Ordering by count ascending

> Ordering by count ascending has unbounded error, and elasticsearch is unable to
> accurately report the error in the doc counts.  We strongly recommend not doing
> this.  If you really must, this is how you sort on doc count ascending:

Reviewer 2: ```suggestion
Ordering terms by ascending document `_count` produces an unbounded error that
{es} can't accurately report. We therefore strongly recommend against using
`"order": { "_count": "asc" }` as shown in the following example:
```
